### PR TITLE
Buffered messages in Ember

### PR DESCRIPTION
**Pull request by RaasAhsan**
Goals of this PR:
1. Considerably simplify how parsers work. In particular, we can avoid reentrancy by reading the HTTP message up until `\r\n\r\n` and then parse the entire message together. The memory overhead remains linear here as before.
2. Handle max header/prelude sizes more clearly.
3. Make the first step towards using more functional parsers. Since we're parsing the entire HTTP message at once, we can theoretically describe a parser for request/response messages and defer to that altogether. This is complicated by the fact that series/0.21 is still on parboiled.
4. Perhaps improve performance a bit. I haven't ran any benchmarks yet, but I would imagine that we get some wins by eliminating reentrancy.

TODO:
- [x] Tests
- [x] More error types

### CODE REVIEW

File: ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala
Comment (RaasAhsan) [inline ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala:32]: I think we shouldn't conflate `maxHeaderSize` (the max bytes of the trailer header section) and with what `maxChunkHeaderSize` is (the max size for chunked bodies?). I'm thinking we should probably bubble the latter out into a separate parameter. Maybe even a separate one for trailers?

File: ember-core/src/main/scala/org/http4s/ember/core/Parser.scala
Comment (RaasAhsan) [inline ember-core/src/main/scala/org/http4s/ember/core/Parser.scala:40]: I think this can be made better in a followup. We don't need to linearly scan the entire array each time we get a new chunk, we can start from 3 bytes before the existing buffer.

File: ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala
Comment (rossabaker) [reply ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala:32]: A separate one for trailers sounds good.  I'm still confused by maxChunkHeaderSize.

File: ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala
Comment (rossabaker) [inline ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala:109]: This is going to be common, right?  Is it worth caching an empty value of this?

File: ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala
Comment (RaasAhsan) [reply ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala:32]: So it looks like `maxChunkHeaderSize` is the max size of the chunk header for each chunk. so `4\r\n` has a length of 3.  Is it worth having a separate configuration parameter for that? Right now it's set to `maxHeaderSize`

Comment (RaasAhsan) [reply ember-core/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala:32]: I'm going to hold off on adding these for now... We're slowly reaching the limit of class parameters we can take (we're at like 15 or 16 right now) so I want to group things together a bit to make it nicer

**Comment by rossabaker** [comment]
@ChristopherDavenport knows this best, but I'll take another look if he falls into a hole.

**Comment by RaasAhsan** [comment]
Definitely want to get Chris' thoughts on this before merging since it's mostly an attempt to make things simpler (though it fixes several bugs)

File: ember-core/src/main/scala/org/http4s/ember/core/Parser.scala
Comment (rossabaker) [inline ember-core/src/main/scala/org/http4s/ember/core/Parser.scala:141]: Is there any additional context we might provide here?

File: ember-core/src/main/scala/org/http4s/ember/core/Parser.scala
Comment (rossabaker) [inline ember-core/src/main/scala/org/http4s/ember/core/Parser.scala:397]: I think this is maybe the third time I've seen these defined, but I guess it's not like they're big or going to change.

File: ember-core/src/main/scala/org/http4s/ember/core/Parser.scala
Comment (rossabaker) [inline ember-core/src/main/scala/org/http4s/ember/core/Parser.scala:400]: `Throwable`?  Seems like `Exception`.

I'd make all these `RuntimeException`, but that's the old Java programmer in me.

File: ember-core/src/main/scala/org/http4s/ember/core/Parser.scala
Comment (RaasAhsan) [reply ember-core/src/main/scala/org/http4s/ember/core/Parser.scala:397]: Maybe they can be pulled into `Shared`

File: ember-core/src/main/scala/org/http4s/ember/core/Parser.scala
Comment (RaasAhsan) [reply ember-core/src/main/scala/org/http4s/ember/core/Parser.scala]: I added in the partially parsed headers. We could go further and produce an error that reflects what the parsing state was, but I'm more inclined to leave that to a parser library once we switch to that

**Review by rossabaker — APPROVED** [review]
Would still love to get @ChristopherDavenport's feedback, but it LGTM.

**Pull request merged**
